### DISCUSSION TITLE
Efficiently drop multiple nodes / neighbors

### DISCUSSION THREAD

**Discussion by xShields** [graphology/graphology#509] [Q&A]
Hi, I noticed issue https://github.com/graphology/graphology/issues/368 , specifically the "drop node faster" portion. I'm currently working with graphs where we need to trigger removal of large amounts of nodes at once, without just clearing the whole graph. We have the node ids to drop in a set already, but I'm noticing things locking up while having to loop over the dropNode function for every item.

In otherwords, doing this:

```
for (let node of nodeIDs) {
  graph.dropNode(node);
}
```

Do you have any other recommendations for efficiently dropping several nodes (and their neighbors) all at once? If not, is optimization of this kind of task on the roadmap currently?

Thanks for your time

P.S. Also just wanted to say thank you for all the effort you guys have put into graphology / sigma. Over the last few years we've gotten a huge amount of value add from it, and any alternatives would have definitely come up short.

**Comment by Yomguithereal** [comment]
Hi @xShields 

> but I'm noticing things locking up

What do you mean by "locking up"? Do you mean freezing the UI some seconds because the operation is slow?

I don't specifically remember what I meant in #368 and of course there exist ways to speed up node/edge deletion from the graph, but the fact is you won't save up an order of magnitude, as the algorithmic complexity of the task should remain the same. I can just optimize some lookups into hashmaps, nothing more. And since the implementation of such optimizations are usually very costly in lines of code and overall complexity, the tradeoff to add them was not favorable.

Usually, people would rely on asynchronous deletion for this kind of case, if you issue is indeed related to UI freezes. But this also add complexity on your end of course.

**Reply by xShields** [reply]
Gotcha, thanks for the context. Yes by locking up, was referring to the UI freezing while the slow operation finishes.

I don't have much experience separating this kind of operation that's in a library already as something asynchronous, but at first glance it seems I should be looking into using a web worker in some way. However, it also looks like there's not an easy way to break that work out separately, since I can't pass the whole graphology object to a worker, and the dropNode function requires that main graph object. 

My first idea was to...

- Use export() to pass a copy of the node / edge information from the main graph instance, and the list of nodes to remove, to the worker
- in worker, make a temporary graph instance, run import() to pass the nodes / edges over, run the calculations / operations of dropping nodes (and associated edges) 
- return the subgraph information to main thread w/ export()
- Call clear() and then import() to replace the node / edge data in the main graph instance as fast as possible

Does this seem like a reasonable solution? It obviously uses more resources to duplicate the graph data. If this is the wrong direction, any tips would be greatly appreciated.

**Reply by Yomguithereal** [reply]
Webworker is not the way to go here. It does not make sense since you will need to serialize & copy the graph to send it back and forth to the webworker. What you need to do is "temporal" asynchronicity here. What I mean by that is that you must drop a chunk of nodes from the graph during the current frame, then wait a little bit using `requestAnimationFrame` or `setTimeout` and drop the next chunk of nodes, etc. until you have dropped all the nodes you wanted to drop in the first place. By ensuring each frame of JavaScript execution does not perform too much work you will avoid the UI from freezing. But this basically makes dropping your nodes an async operations and it can take a while to get used to this style of coding.
